api/agent/runner_client.go

```diff
@@ -346,12 +346,24 @@ func recordFinishStats(ctx context.Context, msg *pb.CallFinished, c pool.RunnerC
 	}
 }
 
+func cloneHeaders(src http.Header) http.Header {
+	dst := make(http.Header, len(src))
+	for k, vs := range src {
+		for _, v := range vs {
+			dst.Add(k, v)
+		}
+	}
+	return dst
+}
+
 func receiveFromRunner(ctx context.Context, protocolClient pb.RunnerProtocol_EngageClient, runnerAddress string, c pool.RunnerCall, done chan error) {
 	w := c.ResponseWriter()
 	defer close(done)
 
 	log := common.Logger(ctx).WithField("runner_addr", runnerAddress)
 	statusCode := int32(0)
+	// Make a copy of header to avoid concurrent read/write error when logCallFinish runs.
+	clonedHeaders := cloneHeaders(w.Header())
```

Comment:
cloning w.Header() is nice, to cover cases where somewhere we already written some headers, but most importantly, we need headers (see line 386) which we receive from runner. In that for-loop, let's add/set headers as well.

Reply:
done